packages/core/core/src/AssetGraph.js

```diff
@@ -187,100 +206,324 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
+
+    // TODO do we really want this?:
+    // one module does `export * from './esm.js'`.
+    // then esm.js has `import something from 'commonjs'`, esm.js doesn’t have any used symbols and isn’t included.
+    // but commonjs was still getting included.
+    // if (sideEffects !== false) return false;
+    // if (dependency.usedSymbols.size === 0) return true;
+    // let parentAsset =
+    //   dependency.value.sourceAssetId != null &&
+    //   this.getNode(dependency.value.sourceAssetId);
+    // if (parentAsset) {
+    //   invariant(parentAsset.type === 'asset');
+    //   if (
+    //     parentAsset.value.symbols != null &&
+    //     parentAsset.value.symbols.size > 0 &&
+    //     parentAsset.usedSymbols.size === 0
+    //   ) {
+    //     return this.getIncomingDependencies(parentAsset.value).every(d => {
+    //       let n = this.getNode(d.id);
+    //       invariant(n && n.type === 'dependency');
+    //       return n.usedSymbols.size === 0;
+    //     });
+    //   }
+    //   return false;
+    // }
+    // return false;
   }
 
-  unmarkParentsWithHasDeferred(node: DependencyNode) {
-    this.traverseAncestors(node, (_node, ctx, actions) => {
-      if (_node.type === 'asset') {
-        let hasDeferred = this.getNodesConnectedFrom(_node).some(_childNode =>
-          _childNode.hasDeferred == null ? false : _childNode.hasDeferred,
-        );
-        if (!hasDeferred) {
-          delete _node.hasDeferred;
+  setUsedSymbolsAssetAddDependency(
+    changedIncomingDeps: $ReadOnlyArray<DependencyNode>,
+    assetNode: AssetNode,
+    outgoingDeps: $ReadOnlyArray<DependencyNode>,
+  ) {
+    // exportSymbol -> identifier
+    let assetSymbols = assetNode.value.symbols;
+    // identifier -> exportSymbol
+    let assetSymbolsInverse = assetNode.value.symbols
+      ? new Map(
+          [...assetNode.value.symbols].map(([key, val]) => [val.local, key]),
+        )
+      : null;
+
+    let hasNamespaceOutgoingDeps = outgoingDeps.some(
+      d => d.value.symbols.get('*')?.local === '*',
+    );
+
+    // 1) Determine what the changedIncomingDeps requests from the asset
+
+    let isEntry = false;
+    // Used symbols that are exported or reexported (symbol will be removed again later) by asset.
+    let assetUsedSymbols: DefaultMap<string, Set<?string>> = new DefaultMap(
+      DefaultMapSet,
+    );
+    // Symbols that have to be namespace reexported by outgoingDeps.
+    let namespaceReexportedSymbols: DefaultMap<
+      string,
+      Set<?string>,
+    > = new DefaultMap(DefaultMapSet);
+
+    if (changedIncomingDeps.length === 0) {
+      // Root in the runtimes Graph
+      assetUsedSymbols.get('*').add(null);
+      namespaceReexportedSymbols.get('*').add(null);
+    } else {
+      for (let incomingDep of changedIncomingDeps) {
+        // TODO isIsolated?
+        if (incomingDep.value.isEntry || incomingDep.value.isIsolated) {
+          isEntry = true;
```

Comment:
The toplevel asset is special because it has no incoming dependencies with symbols and so it would simply be deferred. Not sure if `isEntry || isIsolated` is correct here (JS entry files, JS script tags)